core/restore.php

```diff
@@ -0,0 +1,444 @@
+        $form['backup'][$backup_directory]['#disabled'] = TRUE;
+        $form['backup'][$backup_directory]['#description'] = st('This backup is missing a backup information file and cannot be restored.');
+      }
+      else {
+        $form['backup'][$backup_directory]['#description'] = st('Contains backup files: @list', array(
+          '@list' => implode(', ', array_keys($backup_info['backups'])),
+        ));
+      }
+    }
+
+    // Set the default to the most recent backup.
+    $first_backup = key($form['backup']['#options']);
+    $form['backup']['#default_value'] = $first_backup;
+    $form['backup']['#options'][$first_backup] .= ' ' . st('(most recent)');
+
+    $form['actions'] = array('#type' => 'actions');
+    $form['actions']['submit'] = array(
+      '#type' => 'submit',
+      '#value' => st('Restore backup'),
+    );
+  }
+
+  $form['actions']['cancel'] = array(
+    '#type' => 'link',
+    '#href' => '<front>',
+    '#title' => st('Cancel'),
+  );
+
+  return $form;
+}
+
+/**
+ * Show a progress message while processing a restore.
+ *
+ * Note this page has an HTTP "Refresh" header set, so it's immediately
+ * attempting to load the restore page while being shown.
+ */
+function restore_progress_page($one_time_token) {
+  restore_task_list('restore');
+  backdrop_set_title('Restoring backup');
+
+  // Redirect to restore process page. This uses a one-time token that is
+  // only valid for a few seconds.
+
+  $redirect_path = backdrop_current_script_url(array(
+    'op' => 'restore',
+    'token' => $one_time_token,
+    'time' => REQUEST_TIME,
+    'backup' => $_POST['backup'],
+  ));
+  backdrop_add_http_header('Refresh', '0; ' . $redirect_path);
+
+  $output = '<p>';
+  $output .= st('This process may take a while. Please wait...');
+  $output .= '<span class="restore-progress">&nbsp;</span>';
+  $output .= '</p>';
+  return $output;
+}
+
+/**
+ * Renders a 403 access denied page for restore.php.
+ *
+ * @return string
+ *   Rendered HTML warning with 403 status.
+ */
+function restore_access_denied_page() {
+  backdrop_add_http_header('Status', '403 Forbidden');
+  watchdog('access denied', 'restore.php', NULL, WATCHDOG_WARNING);
+  backdrop_set_title(st('Access denied'));
+
+  $output = '';
+  $steps = array();
+
+  $output .= st('You are not authorized to access this page. Log in using either an account with the <em>restore site backups</em> permission, or the site maintenance account (the account you created during installation). If you cannot log in, you will have to edit <code>settings.php</code> to bypass this access check. To do this:');
+  $output = '<p>' . $output . '</p>';
+
+  $steps[] = st('Find the <code>settings.php</code> file on your system, and open it with a text editor.');
+  $steps[] = st('There is a line inside your <code>settings.php</code> file that says <code>$settings[\'restore_free_access\'] = FALSE</code>. Change it to <code>$settings[\'restore_free_access\'] = TRUE</code>.');
```

Comment:
Can we add a semicolon to the code to copy-paste?
`<code>$settings[\'restore_free_access\'] = TRUE;</code>`
So if people actually paste that, they don't end up with a syntax error.

Maybe off topic: As a dedicated user of settings.**local.**php I find the wording a bit... hmmm. But wording... :shrug:

Reply:
Added the semicolon. The use of `t()`/`st()` needs to be looked at a bit more here.